src/utils/daos_hdlr.c

```diff
@@ -1900,99 +1902,49 @@ parse_filename_dfs(const char *path, char **_obj_name, char **_cont_name)
 		D_STRNDUP(*_cont_name, cont_name,
 			  strlen(cont_name) + 1);
 		if (*_cont_name == NULL)
-			D_GOTO(out, rc = ENOMEM);
+			D_GOTO(out, rc = -DER_NOMEM);
 	}
 	D_STRNDUP(*_obj_name, fname, strlen(fname) + 1);
 	if (*_obj_name == NULL) {
 		D_FREE(*_cont_name);
-		D_GOTO(out, rc = ENOMEM);
+		D_GOTO(out, rc = -DER_NOMEM);
 	}
 out:
 	D_FREE(f1);
 	D_FREE(f2);
 	return rc;
 }
 
-static ssize_t
-write_dfs(struct cmd_args_s *ap, struct file_dfs *file_dfs,
-	  const char *file, void *buf, ssize_t size)
-{
-	int		rc;
-	d_iov_t		iov;
-	d_sg_list_t	sgl;
-
-	sgl.sg_nr = 1;
-	sgl.sg_iovs = &iov;
-	sgl.sg_nr_out = 1;
-	d_iov_set(&iov, buf, size);
-	rc = dfs_write(file_dfs->dfs, file_dfs->obj, &sgl,
-		       file_dfs->offset, NULL);
-	if (rc) {
-		DH_PERROR_SYS(ap, rc, "dfs_write '%s' failed", file);
-		errno = rc;
-		size = -1;
-		D_GOTO(out, rc);
-	}
-	file_dfs->offset += (daos_off_t)size;
-out:
-	return (ssize_t)size;
-}
-
 static ssize_t
 file_write(struct cmd_args_s *ap, struct file_dfs *file_dfs,
 	   const char *file, void *buf, size_t size)
 {
-	ssize_t num_bytes_written = 0;
+	ssize_t		num_bytes_written = 0;
+	int		rc = 0;
+	daos_size_t	write_size;
 
 	if (file_dfs->type == POSIX) {
 		num_bytes_written = write(file_dfs->fd, buf, size);
 	} else if (file_dfs->type == DAOS) {
-		num_bytes_written = write_dfs(ap, file_dfs, file, buf, size);
+		write_size = size;
+		rc = dfs_sys_write(file_dfs->dfs_sys, file_dfs->obj, buf, file_dfs->offset,
+				   &write_size, NULL);
+		if (rc == 0)
+			/* update file pointer with number of bytes written */
+			file_dfs->offset += write_size;
+		num_bytes_written = write_size;
 	} else {
-		fprintf(ap->errstream, "File type not known: %s type=%d\n",
-			file, file_dfs->type);
+		rc = EINVAL;
+		DH_PERROR_SYS(ap, rc, "File type not known '%s' type=%d", file, file_dfs->type);
```

Comment:
Maybe this should also return -1 for num_bytes_written. Otherwise the caller doesn't know that this call failed

Reply:
@daltonbohning I thought dfs_sys_write returned -1 on failure? For the POSIX case write will return -1 on error.

Reply:
But this is the case where an incorrect file type was passed. -1 is not returned in this case.

Reply:
@daltonbohning yeah, but I think maybe in that case I should check for EINVAL after the functional is called? Since it is different than read/write actually failing.

Reply:
@dsikich This is the full function: https://github.com/daos-stack/daos/blob/b447b409375e2a8f193ccdc1868134c3a564160b/src/utils/daos_hdlr.c#L1918-L1946
The problem is that num_bytes_written is returned, not rc. So rc can't be checked. This is part of why I suggested (different comment) changing these functions to have signatures similar to the dfs / dfs_sys signatures, where rc is always returned and something like num_bytes_written is passed/returned by reference.

Reply:
@daltonbohning hmm yeah I think for the wrappers that would be better. I will change these to return rc and fill in num_bytes_written or bytes read.